src/section.rs

```diff
@@ -48,249 +48,355 @@ impl Section {
         }
+    }
 
-        debug!(
-            "{}: merging {} adults into {}",
-            log::prefix(&self.prefix),
-            node::count_adults(params, self.nodes.values()),
-            log::prefix(&parent),
-        );
+    fn handle_relocate_accept(&mut self, node_name: Name, target: Name) -> Option<Action> {
+        if self.outgoing_relocations.remove(&node_name).is_some() {
+            if let Some(mut node) = self.nodes.remove(&node_name) {
+                node.increment_age();
+                if node.is_elder() {
+                    node.demote();
+                    self.chain.insert(
+                        Block::new(Event::Dead, node_name, node.age()),
+                    );
+                }
 
-        let mut section = Section::new(parent);
-        section.chain = self.chain.clone();
-        section.nodes = mem::replace(&mut self.nodes, HashMap::default());
+                return Some(Action::Send(Message::RelocateCommit { node, target }));
+            }
+        }
 
-        self.state = State::Merging(parent);
+        None
+    }
 
-        vec![Response::Merge(section, self.prefix)]
+    fn handle_relocate_reject(
+        &mut self,
+        params: &Params,
+        node_name: Name,
+        target: Name,
+    ) -> Option<Action> {
+        match self.outgoing_relocations.entry(node_name) {
+            Entry::Occupied(mut entry) => {
+                // Do not retry the relocation during startup or if it would trigger merge.
+                if self.prefix == Prefix::EMPTY ||
+                    node::count_adults(params, self.nodes.values()) <= params.group_size
+                {
+                    debug!(
+                        "{}: cancelling relocation of {} (not beneficial anymore)",
+                        log::prefix(&self.prefix),
+                        log::name(&entry.key())
+                    );
+
+                    entry.remove();
+                    None
+                } else {
+                    // Calculate new relocation target.
+                    let target = Hash::from(target).rehash().into();
+
+                    debug!(
+                        "{}: re-initiating relocation of {} to {}",
+                        log::prefix(&self.prefix),
+                        log::name(&entry.key()),
+                        log::name(&target)
+                    );
+
+                    *entry.get_mut() = target;
+                    Some(Action::Send(Message::RelocateRequest { node_name, target }))
+                }
+            }
+            Entry::Vacant(_) => None,
+        }
     }
 
-    fn handle_relocate(&mut self, params: &Params, node: Node) -> Vec<Response> {
-        if !self.relocating_in_nodes.remove(&node.name()) {
-            return Vec::new();
+    fn handle_relocate_commit(&mut self, params: &Params, node: Node) -> Option<Action> {
+        if self.incoming_relocations.remove(&node.name()).is_none() {
+            error!(
```

Comment:
can we have this as panic as this may indicate a bug in code?

Reply:
:+1: